File: simulation/src/nova_sim_mujoco/src/nova_control_ui.cpp
// 终端交互式 nova 控制：关节、笛卡尔目标、夹爪命令行发布（无 GUI）。
#include <iostream>
#include <limits>
#include <sstream>
#include <string>
#include <vector>

#include "calib_sim_mujoco/msg/arm_pose.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "std_msgs/msg/int32.hpp"
#include "std_msgs/msg/string.hpp"

namespace
{
const std::vector<std::string> kJointOrder = {
  "J1_1_joint", "J1_2_joint", "J1_3_joint", "J1_4_joint", "J1_5_joint", "J1_6_joint", "J1_7_joint", "J1_8_joint",
  "J2_1_joint", "J2_2_joint", "J2_3_joint", "J2_4_joint", "J2_5_joint", "J2_6_joint", "J2_7_joint", "J2_8_joint",
  "J3_1_joint", "J3_2_joint", "J3_3_joint", "J3_4_joint", "J3_5_joint", "J3_6_joint",
  "J4_1_joint", "J4_2_joint", "J4_3_joint", "J4_4_joint", "J4_5_joint", "J4_6_joint"};
}  // namespace

/// 简单 CLI：向 /arm_controller/commands、/nova_target_arm_pose、/nova_arm_id、/nova_gripper_goal 发消息。
class NovaControlUiCpp : public rclcpp::Node
{
public:
  NovaControlUiCpp()
  : Node("nova_control_ui_cpp")
  {
    cmd_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>("/arm_controller/commands", 10);
    arm_id_pub_ = create_publisher<std_msgs::msg::Int32>("/nova_arm_id", 10);
    arm_target_pub_ = create_publisher<calib_sim_mujoco::msg::ArmPose>("/nova_target_arm_pose", 10);
    gripper_pub_ = create_publisher<std_msgs::msg::String>("/nova_gripper_goal", 10);
  }

  void run()
  {
    print_help();
    while (rclcpp::ok()) {
      std::cout << "\n[UI] select mode: 1) all joints  2) pose  3) gripper  q) quit\n> " << std::flush;
      std::string cmd;
      if (!std::getline(std::cin, cmd)) {
        break;
      }
      if (cmd == "q" || cmd == "quit") {
        break;
      }
      if (cmd == "1") {
        handle_joint_mode();
      } else if (cmd == "2") {
        handle_pose_mode();
      } else if (cmd == "3") {
        handle_gripper_mode();
      } else {
        std::cout << "[UI] unknown command\n";
      }
    }
  }

private:
  void print_help() const
  {
    std::cout << "Nova Control UI (C++)\n";
    std::cout << "- Joint mode: publish 28 values to /arm_controller/commands\n";
    std::cout << "- Pose mode: publish calib_sim_mujoco/ArmPose to /nova_target_arm_pose\n";
    std::cout << "- Gripper mode: publish /nova_arm_id + /nova_gripper_goal\n";
  }

  bool parse_doubles_line(const std::string & line, std::vector<double> & out) const
  {
    std::stringstream ss(line);
    out.clear();
    double v = 0.0;
    while (ss >> v) {
      out.push_back(v);
    }
    return !out.empty();
  }

  void handle_joint_mode()
  {
    std::cout << "[UI] input 28 joint values separated by spaces.\n";
    std::cout << "[UI] shortcut: input 'zero' for all zeros.\n> " << std::flush;
    std::string line;
    if (!std::getline(std::cin, line)) {
      return;
    }

    std_msgs::msg::Float64MultiArray msg;
    if (line == "zero") {
      msg.data.assign(kJointOrder.size(), 0.0);
    } else {
      std::vector<double> vals;
      if (!parse_doubles_line(line, vals) || vals.size() != kJointOrder.size()) {
        std::cout << "[UI] need exactly 28 values, got " << vals.size() << "\n";
        return;
      }
      msg.data = vals;
    }
    cmd_pub_->publish(msg);
    std::cout << "[UI] published /arm_controller/commands\n";
  }

  void handle_pose_mode()
  {
    std::cout << "[UI] arm_id (0/1/2/3): " << std::flush;
    int arm_id = 0;
    if (!(std::cin >> arm_id)) {
      clear_stdin();
      std::cout << "[UI] invalid arm_id\n";
      return;
    }
    if (arm_id < 0 || arm_id > 3) {
      clear_stdin();
      std::cout << "[UI] arm_id out of range\n";
      return;
    }

    std::cout << "[UI] input x y z qx qy qz qw: " << std::flush;
    double x, y, z, qx, qy, qz, qw;
    if (!(std::cin >> x >> y >> z >> qx >> qy >> qz >> qw)) {
      clear_stdin();
      std::cout << "[UI] invalid pose numbers\n";
      return;
    }
    clear_stdin();

    calib_sim_mujoco::msg::ArmPose arm_pose;
    arm_pose.arm_id = arm_id;
    arm_pose.pose.header.frame_id = "base_link";
    arm_pose.pose.header.stamp = now();
    arm_pose.pose.pose.position.x = x;
    arm_pose.pose.pose.position.y = y;
    arm_pose.pose.pose.position.z = z;
    arm_pose.pose.pose.orientation.x = qx;
    arm_pose.pose.pose.orientation.y = qy;
    arm_pose.pose.pose.orientation.z = qz;
    arm_pose.pose.pose.orientation.w = qw;
    arm_target_pub_->publish(arm_pose);

    std::cout << "[UI] published /nova_target_arm_pose for arm_id=" << arm_id << "\n";
  }

  void handle_gripper_mode()
  {
    std::cout << "[UI] arm_id (0/1): " << std::flush;
    int arm_id = 0;
    if (!(std::cin >> arm_id)) {
      clear_stdin();
      std::cout << "[UI] invalid arm_id\n";
      return;
    }
    if (arm_id < 0 || arm_id > 1) {
      clear_stdin();
      std::cout << "[UI] gripper only supports arm_id 0/1\n";
      return;
    }

    std::cout << "[UI] mode: open / close / width\n> " << std::flush;
    clear_stdin();
    std::string mode;
    if (!std::getline(std::cin, mode)) {
      return;
    }

    std_msgs::msg::Int32 id_msg;
    id_msg.data = arm_id;
    arm_id_pub_->publish(id_msg);

    std_msgs::msg::String gmsg;
    if (mode == "open" || mode == "close") {
      gmsg.data = mode;
    } else if (mode == "width") {
      std::cout << "[UI] width in meters (0~0.1): " << std::flush;
      double width = 0.05;
      if (!(std::cin >> width)) {
        clear_stdin();
        std::cout << "[UI] invalid width\n";
        return;
      }
      clear_stdin();
      gmsg.data = "width:" + std::to_string(width);
    } else {
      std::cout << "[UI] unsupported mode\n";
      return;
    }
    gripper_pub_->publish(gmsg);
    std::cout << "[UI] published gripper command for arm_id=" << arm_id << "\n";
  }

  void clear_stdin()
  {
    std::cin.clear();
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
  }

  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr cmd_pub_;
  rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr arm_id_pub_;
  rclcpp::Publisher<calib_sim_mujoco::msg::ArmPose>::SharedPtr arm_target_pub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr gripper_pub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NovaControlUiCpp>();
  node->run();
  rclcpp::shutdown();
  return 0;
}

File: simulation/src/nova_sim_mujoco/scripts/nova_control_ui.py
#!/usr/bin/env python3
"""
Nova control UI (Tkinter):
1) Joint mode: publish all joints to /arm_controller/commands
2) Pose mode: publish calib_sim_mujoco/ArmPose to /nova_target_arm_pose (与 arm_id 同帧，避免分话题错序)
"""

import tkinter as tk
from tkinter import ttk, messagebox

import rclpy
from calib_sim_mujoco.msg import ArmPose
from std_msgs.msg import Float64MultiArray, Int32, String


JOINT_ORDER = [
    "J1_1_joint", "J1_2_joint", "J1_3_joint", "J1_4_joint", "J1_5_joint", "J1_6_joint", "J1_7_joint", "J1_8_joint",
    "J2_1_joint", "J2_2_joint", "J2_3_joint", "J2_4_joint", "J2_5_joint", "J2_6_joint", "J2_7_joint", "J2_8_joint",
    "J3_1_joint", "J3_2_joint", "J3_3_joint", "J3_4_joint", "J3_5_joint", "J3_6_joint",
    "J4_1_joint", "J4_2_joint", "J4_3_joint", "J4_4_joint", "J4_5_joint", "J4_6_joint",
]


class NovaControlUI:
    def __init__(self) -> None:
        rclpy.init()
        self.node = rclpy.create_node("nova_control_ui")
        self.cmd_pub = self.node.create_publisher(Float64MultiArray, "/arm_controller/commands", 10)
        self.arm_id_pub = self.node.create_publisher(Int32, "/nova_arm_id", 10)
        self.arm_pose_pub = self.node.create_publisher(ArmPose, "/nova_target_arm_pose", 10)
        self.gripper_pub = self.node.create_publisher(String, "/nova_gripper_goal", 10)

        self.root = tk.Tk()
        self.root.title("Nova Arm Control UI")
        self.root.geometry("980x760")
        self.root.protocol("WM_DELETE_WINDOW", self.on_close)

        notebook = ttk.Notebook(self.root)
        notebook.pack(fill=tk.BOTH, expand=True, padx=8, pady=8)

        self.joint_tab = ttk.Frame(notebook)
        self.pose_tab = ttk.Frame(notebook)
        notebook.add(self.joint_tab, text="Joint Control")
        notebook.add(self.pose_tab, text="Pose Control")

        self.joint_vars = []
        self.build_joint_tab()
        self.build_pose_tab()

    def build_joint_tab(self) -> None:
        hdr = ttk.Frame(self.joint_tab)
        hdr.pack(fill=tk.X, padx=8, pady=6)
        ttk.Label(hdr, text="All-joint mode: publish 28 positions to /arm_controller/commands").pack(side=tk.LEFT)

        canvas = tk.Canvas(self.joint_tab)
        scrollbar = ttk.Scrollbar(self.joint_tab, orient="vertical", command=canvas.yview)
        frame = ttk.Frame(canvas)
        frame.bind("<Configure>", lambda e: canvas.configure(scrollregion=canvas.bbox("all")))
        canvas.create_window((0, 0), window=frame, anchor="nw")
        canvas.configure(yscrollcommand=scrollbar.set)
        canvas.pack(side=tk.LEFT, fill=tk.BOTH, expand=True, padx=(8, 0), pady=6)
        scrollbar.pack(side=tk.RIGHT, fill=tk.Y, padx=(0, 8), pady=6)

        for idx, name in enumerate(JOINT_ORDER):
            row = ttk.Frame(frame)
            row.grid(row=idx, column=0, sticky="ew", padx=4, pady=2)
            ttk.Label(row, text=f"{idx:02d} {name}", width=24).pack(side=tk.LEFT)
            v = tk.StringVar(value="0.0")
            self.joint_vars.append(v)
            ttk.Entry(row, textvariable=v, width=12).pack(side=tk.LEFT, padx=4)
            # quick buttons
            ttk.Button(row, text="+0.1", command=lambda i=idx: self.bump_joint(i, +0.1)).pack(side=tk.LEFT, padx=2)
            ttk.Button(row, text="-0.1", command=lambda i=idx: self.bump_joint(i, -0.1)).pack(side=tk.LEFT, padx=2)

        btns = ttk.Frame(self.joint_tab)
        btns.pack(fill=tk.X, padx=8, pady=8)
        ttk.Button(btns, text="Publish Once", command=self.publish_joint_command).pack(side=tk.LEFT, padx=4)
        ttk.Button(btns, text="Set All Zero", command=self.set_all_zero).pack(side=tk.LEFT, padx=4)

    def build_pose_tab(self) -> None:
        body = ttk.Frame(self.pose_tab)
        body.pack(fill=tk.BOTH, expand=True, padx=12, pady=12)

        self.arm_id_var = tk.StringVar(value="0")
        self.frame_id_var = tk.StringVar(value="base_link")
        self.px_var = tk.StringVar(value="0.35")
        self.py_var = tk.StringVar(value="-0.10")
        self.pz_var = tk.StringVar(value="0.25")
        self.qx_var = tk.StringVar(value="0.0")
        self.qy_var = tk.StringVar(value="0.0")
        self.qz_var = tk.StringVar(value="0.0")
        self.qw_var = tk.StringVar(value="1.0")

        self.gripper_mode_var = tk.StringVar(value="none")
        self.gripper_width_var = tk.StringVar(value="0.05")

        row = 0
        ttk.Label(body, text="Pose mode: /nova_target_arm_pose (ArmPose)").grid(row=row, column=0, columnspan=4, sticky="w")
        row += 1

        ttk.Label(body, text="arm_id (0/1/2/3)").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        ttk.Entry(body, textvariable=self.arm_id_var, width=12).grid(row=row, column=1, sticky="w")
        ttk.Label(body, text="frame_id").grid(row=row, column=2, sticky="e", padx=6)
        ttk.Entry(body, textvariable=self.frame_id_var, width=16).grid(row=row, column=3, sticky="w")
        row += 1

        ttk.Label(body, text="x").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        ttk.Entry(body, textvariable=self.px_var, width=12).grid(row=row, column=1, sticky="w")
        ttk.Label(body, text="y").grid(row=row, column=2, sticky="e", padx=6)
        ttk.Entry(body, textvariable=self.py_var, width=12).grid(row=row, column=3, sticky="w")
        row += 1
        ttk.Label(body, text="z").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        ttk.Entry(body, textvariable=self.pz_var, width=12).grid(row=row, column=1, sticky="w")
        row += 1

        ttk.Label(body, text="qx").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        ttk.Entry(body, textvariable=self.qx_var, width=12).grid(row=row, column=1, sticky="w")
        ttk.Label(body, text="qy").grid(row=row, column=2, sticky="e", padx=6)
        ttk.Entry(body, textvariable=self.qy_var, width=12).grid(row=row, column=3, sticky="w")
        row += 1
        ttk.Label(body, text="qz").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        ttk.Entry(body, textvariable=self.qz_var, width=12).grid(row=row, column=1, sticky="w")
        ttk.Label(body, text="qw").grid(row=row, column=2, sticky="e", padx=6)
        ttk.Entry(body, textvariable=self.qw_var, width=12).grid(row=row, column=3, sticky="w")
        row += 1

        ttk.Separator(body, orient=tk.HORIZONTAL).grid(row=row, column=0, columnspan=4, sticky="ew", pady=8)
        row += 1

        ttk.Label(body, text="Gripper").grid(row=row, column=0, sticky="e", padx=6, pady=4)
        mode_combo = ttk.Combobox(
            body,
            textvariable=self.gripper_mode_var,
            values=["none", "open", "close", "width"],
            state="readonly",
            width=10,
        )
        mode_combo.grid(row=row, column=1, sticky="w")
        ttk.Label(body, text="width(m)").grid(row=row, column=2, sticky="e", padx=6)
        ttk.Entry(body, textvariable=self.gripper_width_var, width=12).grid(row=row, column=3, sticky="w")
        row += 1

        btns = ttk.Frame(body)
        btns.grid(row=row, column=0, columnspan=4, sticky="w", pady=12)
        ttk.Button(btns, text="Send Pose", command=self.send_pose_goal).pack(side=tk.LEFT, padx=4)
        ttk.Button(btns, text="Send Gripper Only", command=self.send_gripper_only).pack(side=tk.LEFT, padx=4)

    def bump_joint(self, idx: int, delta: float) -> None:
        try:
            v = float(self.joint_vars[idx].get())
        except ValueError:
            v = 0.0
        self.joint_vars[idx].set(f"{v + delta:.4f}")

    def set_all_zero(self) -> None:
        for v in self.joint_vars:
            v.set("0.0")

    def publish_joint_command(self) -> None:
        try:
            values = [float(v.get()) for v in self.joint_vars]
        except ValueError:
            messagebox.showerror("Input Error", "Joint value must be numeric.")
            return
        msg = Float64MultiArray()
        msg.data = values
        self.cmd_pub.publish(msg)
        self.node.get_logger().info("Published /arm_controller/commands")

    def send_pose_goal(self) -> None:
        try:
            arm_id = int(self.arm_id_var.get())
            px = float(self.px_var.get())
            py = float(self.py_var.get())
            pz = float(self.pz_var.get())
            qx = float(self.qx_var.get())
            qy = float(self.qy_var.get())
            qz = float(self.qz_var.get())
            qw = float(self.qw_var.get())
        except ValueError:
            messagebox.showerror("Input Error", "arm_id / pose values must be numeric.")
            return
        if arm_id not in (0, 1, 2, 3):
            messagebox.showerror("Input Error", "arm_id must be 0,1,2,3.")
            return

        out = ArmPose()
        out.arm_id = arm_id
        out.pose.header.frame_id = self.frame_id_var.get().strip() or "base_link"
        out.pose.header.stamp = self.node.get_clock().now().to_msg()
        out.pose.pose.position.x = px
        out.pose.pose.position.y = py
        out.pose.pose.position.z = pz
        out.pose.pose.orientation.x = qx
        out.pose.pose.orientation.y = qy
        out.pose.pose.orientation.z = qz
        out.pose.pose.orientation.w = qw
        self.arm_pose_pub.publish(out)

        mode = self.gripper_mode_var.get()
        if mode != "none":
            self.publish_gripper_mode(mode)
        self.node.get_logger().info(f"Sent pose goal for arm_id={arm_id}")

    def publish_gripper_mode(self, mode: str) -> None:
        msg = String()
        if mode == "width":
            try:
                width = float(self.gripper_width_var.get())
            except ValueError:
                messagebox.showerror("Input Error", "gripper width must be numeric.")
                return
            msg.data = f"width:{width:.4f}"
        else:
            msg.data = mode
        self.gripper_pub.publish(msg)

    def send_gripper_only(self) -> None:
        try:
            arm_id = int(self.arm_id_var.get())
        except ValueError:
            messagebox.showerror("Input Error", "arm_id must be numeric.")
            return
        mode = self.gripper_mode_var.get()
        if mode == "none":
            messagebox.showerror("Input Error", "Please choose gripper mode.")
            return
        id_msg = Int32()
        id_msg.data = arm_id
        self.arm_id_pub.publish(id_msg)
        self.publish_gripper_mode(mode)
        self.node.get_logger().info(f"Sent gripper={mode} for arm_id={arm_id}")

    def on_close(self) -> None:
        try:
            self.node.destroy_node()
            rclpy.shutdown()
        finally:
            self.root.destroy()

    def run(self) -> None:
        self.root.mainloop()


def main() -> None:
    ui = NovaControlUI()
    ui.run()


if __name__ == "__main__":
    main()
